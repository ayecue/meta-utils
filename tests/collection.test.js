const { Container, Signature } = require('../dist');
const GeneralSignatures = require('./mocks/signatures/general.json');
const StringSignatures = require('./mocks/signatures/string.json');
const SubStringSignatures = require('./mocks/signatures/sub-string.json');
const EN = require('./mocks/descriptions/en');

describe('collection', () => {
  let meta = null;

  beforeEach(() => {
    meta = new Container();

    meta.addTypeSignatureFromPayload(GeneralSignatures);
    meta.addTypeSignatureFromPayload(StringSignatures);
    meta.addTypeSignatureFromPayload(SubStringSignatures);

    meta.addMeta('en', EN);
  });

  test('should return signatures', () => {
    expect(meta.getTypeSignature('general').toJSON()).toEqual(Signature.parse(GeneralSignatures).toJSON());
  });

  test('should return description', () => {
    expect(meta.getDefinition('general', 'print').description).toEqual(EN.general.print.description);
  });

  test('should return example', () => {
    expect(meta.getDefinition('general', 'print').example).toEqual(EN.general.print.example);
  });

  test('should return definition', () => {
    expect(meta.getDefinition(['string'], 'split').toJSON()).toMatchSnapshot();
  });

  test('should return property definition', () => {
    expect(meta.getDefinition('sub-string', 'myProperty').toJSON()).toMatchSnapshot();
  });

  test('should return definitions of child definition', () => {
    expect(meta.getDefinition(['sub-string'], 'split').toJSON()).toMatchSnapshot();
  });

  test('should return definitions of parent definition', () => {
    expect(meta.getDefinition(['sub-string'], 'hasIndex').toJSON()).toMatchSnapshot();
  });

  test('should create fork', () => {
    const result = meta.fork();

    expect(result.signatures).toEqual(meta.signatures);
    expect(result.meta).toEqual(meta.meta);
  });

  test('should override fork but keep original', () => {
    const result = meta.fork();

    result.addMeta('en', {
      'general': {
        'print': {
          description: 'test',
          example: ['test']
        }
      }
    });

    expect(result.getDefinition('general', 'print').description).toEqual('test');
    expect(meta.getDefinition('general', 'print').description).toEqual(EN.general.print.description);
  });
});