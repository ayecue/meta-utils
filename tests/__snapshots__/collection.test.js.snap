// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`collection should return definition 1`] = `
Object {
  "arguments": Array [
    Object {
      "default": null,
      "label": "delimiter",
      "opt": false,
      "type": "string",
    },
  ],
  "description": "Returns a \`list\` where each item is a segment of the \`string\`. The separation depends on the provided separator \`string\`. Keep in mind that for some odd reason, this method is using regular expressions for matching.",
  "example": Array [
    "myString = \\"42 as an answer is wrong\\"",
    "segments = myString.split(\\" \\")",
    "if segments[0] != \\"42\\" then",
    "   print(\\"Invalid information spotted!\\")",
    "else",
    "   print(\\"Information seems valid!\\")",
    "end if",
  ],
  "isProtected": false,
  "returns": Array [
    "list:string",
  ],
  "type": "function",
}
`;

exports[`collection should return definitions of parent definition 1`] = `
Object {
  "arguments": Array [
    Object {
      "default": null,
      "label": "delimiter",
      "opt": false,
      "type": "string",
    },
  ],
  "description": "Returns a \`list\` where each item is a segment of the \`string\`. The separation depends on the provided separator \`string\`. Keep in mind that for some odd reason, this method is using regular expressions for matching.",
  "example": Array [
    "myString = \\"42 as an answer is wrong\\"",
    "segments = myString.split(\\" \\")",
    "if segments[0] != \\"42\\" then",
    "   print(\\"Invalid information spotted!\\")",
    "else",
    "   print(\\"Information seems valid!\\")",
    "end if",
  ],
  "isProtected": false,
  "returns": Array [
    "list:string",
  ],
  "type": "function",
}
`;

exports[`collection should return property definition 1`] = `
Object {
  "description": "Test, test!",
  "example": Array [
    "testing",
  ],
  "isProtected": true,
  "type": Object {
    "type": "list",
    "valueType": "string",
  },
}
`;

exports[`collection should return signatures 1`] = `
Object {
  "definitions": Object {
    "hasIndex": Object {
      "arguments": Array [
        Object {
          "default": null,
          "label": "value",
          "opt": false,
          "type": "any",
        },
        Object {
          "default": null,
          "label": "index",
          "opt": false,
          "type": "any",
        },
      ],
      "description": null,
      "example": null,
      "isProtected": false,
      "returns": Array [
        "number",
      ],
      "type": "function",
    },
    "print": Object {
      "arguments": Array [
        Object {
          "default": "\\"\\"",
          "label": "value",
          "opt": false,
          "type": "any",
        },
        Object {
          "default": null,
          "label": "delimiter",
          "opt": true,
          "type": "string",
        },
      ],
      "description": null,
      "example": null,
      "isProtected": false,
      "returns": Array [
        "null",
      ],
      "type": "function",
    },
  },
  "extend": null,
  "type": "general",
}
`;
